File: LeetCode/438找到字符串中所有字母异位词.go
//切片不能用==判等，但是数组可以用==判等
func findAnagrams(s string, p string) []int {
	lens, lenp := len(s), len(p)
	var ans []int
	if lens < lenp {
		return ans
	}
	var chs [26]int
	var chp [26]int
	for i := 0; i < lenp; i++ {
		chs[s[i]-'a']++
		chp[p[i]-'a']++
	}
	if chs == chp {
		ans = append(ans, 0)
	}
	for i := lenp; i < lens; i++ {
		chs[s[i-lenp]-'a']--
		chs[s[i]-'a']++
		if chs == chp {
			ans = append(ans, i-lenp+1)
		}
	}
	return ans
}

File: LeetCode/438找到字符串中所有字母异位词.java
// Arrays.equals表示对比两个数组是否相等
// 滑动窗口
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int lens = s.length(), lenp = p.length();
        if (lens < lenp) return new ArrayList<Integer>();
        List<Integer> ans = new ArrayList<Integer>();
        int[] chsp = new int[26];
        int[] chss = new int[26];
        for (int i = 0; i < lenp; i++) {
            ++chss[s.charAt(i) - 'a'];
            ++chsp[p.charAt(i) - 'a'];
        }
        if (Arrays.equals(chss, chsp)) {
            ans.add(0);
        }
        for (int i = lenp; i < lens; i++) {
            --chss[s.charAt(i - lenp) - 'a'];
            ++chss[s.charAt(i) - 'a'];
            if (Arrays.equals(chss, chsp)) {
                ans.add(i - lenp + 1);
            }
        }
        return ans;
    }
}